pilot/pkg/proxy/envoy/v1/gateway.go

```diff
@@ -0,0 +1,224 @@
+	}
+}
+
+// TODO: this isn't really correct: we need xDS v2 APIs to really configure this correctly.
+// Our TLS options align with SDSv2 DownstreamTlsContext, but the v1 API's SSLContext is split
+// into three pieces; we need at least two of the pieces here.
+func tlsToSSLContext(tls *routing.Server_TLSOptions, protocol string) *SSLContext {
+	return &SSLContext{
+		CertChainFile:            tls.ServerCertificate,
+		PrivateKeyFile:           tls.PrivateKey,
+		CaCertFile:               tls.CaCertificates,
+		RequireClientCertificate: tls.Mode == routing.Server_TLSOptions_MUTUAL,
+		ALPNProtocols:            strings.Join(ListenersALPNProtocols, ","),
+	}
+}
+
+// buildGatewayHTTPRoutes creates HTTP route configs for a single external port on a gateway
+func buildGatewayVirtualHosts(configStore model.IstioConfigStore, node model.Node, listenerPort int) (*HTTPRouteConfig, error) {
+
+	gateways, err := configStore.List(model.Gateway.Type, model.NamespaceAll)
+	if err != nil {
+		return nil, err
+	}
+
+	allRules, err := configStore.List(model.V1alpha2RouteRule.Type, model.NamespaceAll)
+	if err != nil {
+		return nil, fmt.Errorf("getting all rules: %s", err)
+	}
+
+	virtualHosts := []*VirtualHost{}
+	for _, gwConfig := range gateways {
+		gatewayName := gwConfig.Name
+		gateway := gwConfig.Spec.(*routing.Gateway)
+		for _, server := range gateway.Servers {
+			if listenerPort != int(server.Port.Number) {
+				continue
+			}
+			for _, externalHostname := range server.Hosts {
+				routesForThisVirtualHost, err := buildDestinationHTTPRoutesForGatewayVirtualHost(allRules,
+					node, configStore, listenerPort, buildOutboundCluster,
+					gatewayName, externalHostname)
+				if err != nil {
+					return nil, err
+				}
+				if len(routesForThisVirtualHost) == 0 {
+					log.Warn("gateway-virtual-hosts", zap.String("host", externalHostname))
+					continue
+				}
+
+				virtualHosts = append(virtualHosts, &VirtualHost{
+					Name:    fmt.Sprintf("%s|%d|%s", gatewayName, listenerPort, externalHostname),
+					Domains: []string{fmt.Sprintf("%s:%d", externalHostname, listenerPort), externalHostname},
+					Routes:  routesForThisVirtualHost,
+				})
+			}
+		}
+	}
+
+	return &HTTPRouteConfig{
+		VirtualHosts: virtualHosts,
+	}, nil
+}
+
+func buildDestinationHTTPRoutesForGatewayVirtualHost(
+	allRules []model.Config,
+	node model.Node, config model.IstioConfigStore,
+	externalPort int,
+	buildCluster buildClusterFunc,
+	gatewayName string, externalHostname string) ([]*HTTPRoute, error) {
+
+	rules := filterRulesToGatewayAndExternalHostname(allRules, gatewayName, externalHostname)
+	log.Debug("gateway-virtual-hosts", zap.String("host", externalHostname), zap.Int("rules", len(rules)))
+	if len(rules) == 0 {
+		return nil, nil
+	}
+	// len(rules) <= 1 is guaranteed because v1alpha2 rules are unique per host.
+	if len(rules) > 1 {
+		return nil, fmt.Errorf("unexpectedly got %d route rules for %s", len(rules), externalHostname)
+	}
+	rule := rules[0]
+
```

Comment:
Why do you need this in the first place? Why not just assume that func returns a single instance of a rule object ?